### PR TITLE
[Merged by Bors] - syncer: re-sync data from peers with diverged mesh hash

### LINKED ISSUE #2507: mesh hash resolution with peers
# design discussion
https://github.com/spacemeshos/SMIPS/issues/51 
https://github.com/spacemeshos/SMIPS/issues/60

# implementation plan
## fetch
add the following to fetch interface
```
type MeshHashRequest struct {
    From, To     types.LayerID
    Delta, Steps uint32
}

type MeshHashes struct {
    Layers []types.LayerID
    Hashes []types.Hash32
}

type EpochData struct {
    Beacon types.Beacon
    Weight uint64
    AtxIDs []types.ATXID
}

type layerFetcher interface {
    Peers() []p2p.Peer
    PeerEpochInfo(context.Context, p2p.Peer, types.EpochID) (*fetch.EpochData, error)
    PeerMeshHashes(context.Context, p2p.Peer, *fetch.MeshHashRequest) (*fetch.MeshHashes, error)
}
```

## syncer
syncer periodically check peers’ agreement on the mesh hash (aka aggregated hash) along with its layer opinions.

given a list of opinions for layer `N` from peers, syncer first finds those that differ from the node’s mesh hash at layer `N-1`. for those peers. pseudo code below:

```
getMeshAgreement(lid types.LayerID, opinions []*fetch.LayerOpinion) (*fetch.LayerOpinion, error) {
    maxIgnoredWeight uint64
    candidate LayerOpinion
    for opn in opinions {
        if opn.meshHash == nodeMeshHash {
            // update last agreement with peer
            forkFinder.UpdateAgreement(opn.Peer(), lid.Sub(1))
        else {
            epochData = fetch.PeerEpochInfo()
            missing = set_diff(peerAtxIDs, nodeAtxIDs)  // in peerAtxIDs but not in nodeAtxIDs
            err = fetchATX(missing)
            if err {
                // if the node cannot download the ATXs claimed by this peer, 
                // it does not trust this peer's mesh
                continue
            }
            ignoredWeight = total_weight(missing)
            if ignoredWeight > maxIgnoredWeight {
                maxIgnoredWeight = ignoredWeight
                candidate = opn
            }
        }
    }
    
    threshold := (nodeEpochWeight + maxIgnoredWeight).Fraction(s.conf.TortoiseGlobalThreshold)
    if nodeEpochWeight > threshold {
        // the node should still be confident of its mesh
        return nil, nil
    }

    // find the divergent layer and adopt the peer's mesh from there
    fork = forkFinder.FindFork(candidate.Peer(), to, candidate.PrevAggHash)
    adoptPeerMesh(ctx, candidate.Peer(), fork.Add(1), lid.Sub(1))
    // clear the agreement cache after importing new opinions
    s.forkFinder.Purge(true)
    return candidate, nil
}
```

## fork finder
add the following to facilitate finding divergent layer with peers.
```
type layerHash struct {
    lid     types.LayerID
    hash    types.Hash32
    created time.Time
}

type ForkFinder struct {
    logger           log.Log
    db               *sql.Database
    fetcher          layerFetcher
    // max number of hashes requested from peer in a request
    maxHashesInReq   uint32
    // any cached peer hash will be purged after this duration if the peer is no longer the node's peer list
    maxStaleDuration time.Duration

    mu          sync.Mutex
    // caches the last agreed hash with peer
    agreedPeers map[p2p.Peer]*layerHash
}

// FindFork finds the divergent layer with peer between the last agreed layer and the layer where
// difference is detected.
FindFork(peer p2p.Peer, diffLid types.LayerID, diffHash types.Hash32) (types.LayerID, error) {
}

// UpdateAgreement updates the last layer that node and peer agreed on the mesh hash.
UpdateAgreement(peer p2p.Peer, lid types.LayerID, hash types.Hash32) {
}
```
### basic algorithm
- LA: latest agreement layer, fall back to genesis layer if not available
- OD:  oldest disagreement layer
- for each request to peer, make sure the boundary hashes are not changed, as the peer can be going through reorg during the requests
- the number of hashes requested each time is bounded by `maxHashesInReq`. based on that, calculate the delta layers in each step, and the number of steps between LA and OD to get `maxHashesInReq` of hashes in requests.

### example
assuming 
- 3 layers per epoch
- `maxHashesInReq` is 5
- node and peer X forked at layer 35:
- LA: N/A
- OD: 50

the resulting requests with peer X are
- req 0: [5, 19, 33, 47, 50]    -> LA: 33, OD: 47
- req 1: [33, 37, 41, 45, 47]  -> LA: 33, OD: 37
- req 2: [33, 34, 35, 36, 37]  -> LA: 35, OD: 36 -> fork: 35

## adopting a peer's mesh
as per https://github.com/spacemeshos/go-spacemesh/issues/3573, tortoise will be changed to such that as long as tortoise is fed with new data it will change opinion on the previous layers

assuming syncer decides to adopt the mesh from a confident peer from layer `F+1` to `N-1`, it will first fetch the missing ATXs, and ballots/blocks layer by layer from `F+1`.

### data
- for every new ATX received after its publish epoch
  - call tortoise `OnAtx(atx)` handler 
- for every ballot that reference an ignored ATX
  - call tortoise `OnBallot(ballot)` handler
  - set the ballot weight to 0 (similar to an malicious ballot)
- all blocks are fetched and saved as usual (call `OnBlock(block)`)

### opinions
as per https://github.com/spacemeshos/go-spacemesh/issues/3573, for genesis, full tortoise using 10 000 layers as a sliding window is completely viable to run. therefore importing opinions from confident peers are not relevant (importing opinions is only relevant for verifying tortoise). 

as a result, in the event of network rejoin from a 30/70 split, nodes on the 70-split will remain confident. nodes on the 30-split will do full vote counting on ballots in both sides (self-healing without voting by coin flip) and reach the same mesh hash as the 70-split nodes.

### LINKED ISSUE #2527: implement syncer code that drives hash resolution
## Description
part of #2507

implement syncer code that drives hash resolution as described (in pink) in the following chart

sync-cfg

specifically, the box `resolve hash difference` is expanded to the diagram below:

hash-resolution-cfg

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #2507
Closes #2527

## Changes

- syncer: remove code syncing block validity. this is done to simplified genesis logic, as full tortoise is feasible.

- syncer: when detecting peers has a different mesh hash, taking the following measure
  - for each peer, find fork layer with that peer and resync from that peer [fork+1, latest]
  - avoid resync from the same peer within the same run
  - avoid resync based on the same layer hash divergence
  - after all data is collected, trigger ProcessLayer to allow tortoise to tally votes

- fetch: use EpochInfo to capture weight/beacon/active set when peers exchange epoch data. this is used 
  - to sync epoch ATXs
  - to avoid malicious peer forcing the node to do endless fork-finding session, make sure the peer's active set is valid prior to entering a session

- add systests for 30/70 and 50/50 splits. the success condition is that all nodes agreed on the state hash for every layer

### CODE REVIEW

File: systest/tests/partition_test.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	require.NoError(t, finalErr)
+	require.True(t, pass)
+	_ = txeg.Wait()
+}
+
+func TestPartition_30_70(t *testing.T) {
+	t.Parallel()
+
+	tctx := testcontext.New(t, testcontext.Labels("sanity"))
```
Comment (dshulyak) [inline systest/tests/partition_test.go]: we may have to remove it from sanity group. and create a separate job for lets say destructive group. or use more resources and parallize tests better

there were some problems if tests were running for more than 60m

Comment (countvonzero) [reply systest/tests/partition_test.go]: > create a separate job

do you mean to have two systest jobs running concurrently?

Comment (dshulyak) [reply systest/tests/partition_test.go]: one by one. for concurrent we need to provision more clusters at the same time

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/actions/runs/3388705783/jobs/5632226985)

the usual TestAddNodes failure

File: tortoise/tortoise.go [hunk 2]
```diff
@@ -486,9 +483,15 @@ func (t *turtle) verifyLayers() error {
...
-			if err != nil {
-				return fmt.Errorf("saving validity for %s: %w", block.id, err)
+			valid := block.validity == support
+			if valid {
+				if err := blocks.SetValid(t.cdb, block.id); err != nil {
+					return fmt.Errorf("setting %s valid: %w", block.id, err)
+				}
+			} else {
+				if err := blocks.SetInvalid(t.cdb, block.id); err != nil {
+					return fmt.Errorf("setting block %s invalid: %w", block.id, err)
```
Comment (dshulyak) [inline tortoise/tortoise.go]: this wrapping is completely unnecessary, it doesn't add any new information

but i think this change will have to be reverted

**Review by dshulyak — COMMENTED** [review]
revert decision should be based on tortoise counting
if tortoise terminates with X, then all nodes change the vote to Y (note that there is no partition) - they all will remain with X applied

so in order to detect that opinion changed to Y - layer needs to be reverted based on tortoise counting, not on what peer is saying

File: systest/tests/partition_test.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/systest/chaos"
+	"github.com/spacemeshos/go-spacemesh/systest/cluster"
+	"github.com/spacemeshos/go-spacemesh/systest/testcontext"
+)
+
+func sendTransactions(tctx *testcontext.Context, cl *cluster.Cluster, stop uint32) *errgroup.Group {
```
Comment (dshulyak) [inline systest/tests/partition_test.go]: is it possible to reuse previous code for sending transactions?

File: fetch/wire_types.go [hunk 4]
```diff
@@ -52,6 +52,12 @@ type MeshHashes struct {
 	Hashes []types.Hash32
 }
 
+type EpochData struct {
+	Beacon types.Beacon
+	Weight uint64
```
Comment (dshulyak) [inline fetch/wire_types.go]: can we please drop this two fields? both of them should have zero impact on downloading atxs. if they do then something is clearly wrong

**Comment by dshulyak** [comment]
> to determine whether to trust a peer's data prior to entering a fork-finding session with the peer. i.e. if the node fails to fetch the active set from the peer, it will not trust the peer's mesh hash.

i don't understand what is it for. syncer doesn't need to trust peer's mesh hash, it needs to download missing data. if data is valid it will be used by tortoise and thats it

> gather the minFork layer from all peers, and re-sync layer data from them from minFork+1 layer

it makes more sense to individually download diff from every peer based on the fork with that peer

> syncer: prior to adopting a peer's opinions, do some basic validity checking on the opinions

i would consider dropping peers opinions for genesis, and maybe restore them in some distant future. it is an optimization that introduces even more complexity in the syncer

File: syncer/data_fetch.go [hunk 5]
```diff
@@ -51,23 +52,25 @@ type (
...
-		fetcher: fetch,
 	}
 }
 
 // PollLayerData polls all peers for data in the specified layer.
-func (d *DataFetch) PollLayerData(ctx context.Context, lid types.LayerID) error {
-	peers := d.fetcher.GetPeers()
+func (d *DataFetch) PollLayerData(ctx context.Context, lid types.LayerID, peers ...p2p.Peer) error {
+	peers = append(peers, d.fetcher.GetPeers()...)
+	peers = util.UniqueSliceStringer(peers)
```
Comment (dshulyak) [inline syncer/data_fetch.go]: that function should not be used for anything beside logging. peer.ID is already a string, String method copies that data as a []byte buffer and encodes it with b58

File: syncer/data_fetch.go [hunk 6]
```diff
@@ -51,23 +52,25 @@ type (
...
 		msh:     msh,
-		fetcher: fetch,
 	}
 }
 
 // PollLayerData polls all peers for data in the specified layer.
-func (d *DataFetch) PollLayerData(ctx context.Context, lid types.LayerID) error {
-	peers := d.fetcher.GetPeers()
+func (d *DataFetch) PollLayerData(ctx context.Context, lid types.LayerID, peers ...p2p.Peer) error {
+	peers = append(peers, d.fetcher.GetPeers()...)
```
Comment (dshulyak) [inline syncer/data_fetch.go]: where fetcher gets peers? it is from currently open connections? where do you get peers that are sent to this method?

Comment (dshulyak) [reply syncer/data_fetch.go]: they are also from fetcher.. this doesn't makes sense then

**Comment by dshulyak** [comment]
what happens if data was download from peer but local opinion didn't change, it can happen atleast for two reasons:
- ballots with different beacon counted are delayed
- local opinion can be more "correct"

is there something that will prevent querying peer with different opinion again?

File: cmd/node/node_test.go [hunk 7]
```diff
@@ -817,6 +817,12 @@ func TestInitialize_BadTortoiseParams(t *testing.T) {
 	app = New(WithLog(logtest.New(t)), WithConfig(&conf))
 	err := app.Initialize()
 	assert.EqualError(t, err, "incompatible tortoise hare params")
+
+	conf, err = presets.Get("fastnet")
```
Comment (dshulyak) [inline cmd/node/node_test.go]: this seems out of palce

**Comment by countvonzero** [comment]
> > syncer: prior to adopting a peer's opinions, do some basic validity checking on the opinions
> 
> i would consider dropping peers opinions for genesis, and maybe restore them in some distant future. it is an optimization that introduces even more complexity in the syncer

are you saying that a new node doesn't need block validity from peers to bootstrap verifying tortoise local opinion?
see 2nd point on https://github.com/spacemeshos/go-spacemesh/issues/3003

**Comment by countvonzero** [comment]
> what happens if data was download from peer but local opinion didn't change, it can happen atleast for two reasons:
> 
> * ballots with different beacon counted are delayed
> * local opinion can be more "correct"
> 
> is there something that will prevent querying peer with different opinion again?

in the current code in develop, if the block validity is set already it will not query peer for block validity again.
but there is a race condition, hence adding `blocks.SetValidIfNotSet`

**Comment by countvonzero** [comment]
> revert decision should be based on tortoise counting if tortoise terminates with X, then all nodes change the vote to Y (note that there is no partition) - they all will remain with X applied
> 
> so in order to detect that opinion changed to Y - layer needs to be reverted based on tortoise counting, not on what peer is saying

i know this idea is conceptually incorrect, as the mesh content should be based on tortoise's vote counting decision.
but i did it for efficiency, as 

- blocks are optimistically applied. i.e. the validity is always decided after a block is applied that can be a possible revert
- even without partition, nodes can disagree on mesh hash. some node can fail locally with hare and didn't apply anything. 
in this case, it also rely on syncer to detect that there is a hash difference with peers and try to download data from them.

ultimately the fork detection in syncer doesn't care whether there is a partition or not. it just cares that there is disagreement from peers. 

the risk is that when all nodes agree on the mesh content (based on hare output), then nothing will drive the network to the correct tortoise output. i checked the logs on other systests (TestAddNodes and TestFailedNodes), the syncer fork detection occurred many times as well.

in any case, i am happy to restore to the old behavior. tho the simplicity of checking for revert appeal to me. @dshulyak

**Comment by countvonzero** [comment]
> > to determine whether to trust a peer's data prior to entering a fork-finding session with the peer. i.e. if the node fails to fetch the active set from the peer, it will not trust the peer's mesh hash.
> 
> i don't understand what is it for. syncer doesn't need to trust peer's mesh hash, it needs to download missing data. if data is valid it will be used by tortoise and thats it

but how else are we going to prevent malicious peers trying to get the node endlessly trying to sync non-existing data?
the only way i can think of, is to double check if the peer has all the ATXs it claimed it has. we had a discussion in slack a while back, you also asked the same question.

beacon is indeed unnecessary/unused. i only logged the difference now. 

UPD: the approach (verifying peers' active set) is also suggested by Tal when discussing mesh resolution. that was also 1 yr back.

**Comment by dshulyak** [comment]
> are you saying that a new node doesn't need block validity from peers to bootstrap verifying tortoise local opinion?

verifying tortoise is an optimization. we don't have to use it for syncing a node. that was a mistake due to miscommunication  with research

> in any case, i am happy to restore to the old behavior. tho the simplicity of checking for revert appeal to me. 

i think there are just too many problems with that:
1. timing. syncer may download ballots with different beacon that will be delayed. and opinion will change long after syncer reverted the layer where opinion hash was different. e.g. it run revert at point X, 10 layer after tortoise counted more delayed ballots, opinion actually changed, but we didn't revert
2. peers do not commit to the provided layer hash. so they may say that opinion is X, but counting votes will result in Y
3. we currently don't update layer hash every time when tortoise changes vote. so it is not guaranteed that layer hash in mesh will be updated if originally vote was X and then became Y

**Comment by dshulyak** [comment]
> in the current code in develop, if the block validity is set already it will not query peer for block validity again.
but there is a race condition, hence adding blocks.SetValidIfNotSet

my question was about not trying to run fork finder with peer if we already downloaded data from that peer, but opinion didn't change

> but how else are we going to prevent malicious peers trying to get the node endlessly trying to sync non-existing data?

maybe i misunderstood the comment then. clearly if node can't download something from a peer it shouldn't try to download more stuff

**Comment by countvonzero** [comment]
> what happens if data was download from peer but local opinion didn't change, it can happen atleast for two reasons:
> 
> * ballots with different beacon counted are delayed
> * local opinion can be more "correct"
> 
> is there something that will prevent querying peer with different opinion again?

not in the current PR. tho i'll need to remember the disagreed hash across syncer runs.
the thing is that the same peer can also be changing its opinions, so its mesh content can change across syncer runs.
but i can cache the peer+hash combo to avoid another fork-finding session.

File: fetch/wire_types.go (hunk 4 above)
Comment (countvonzero) [reply fetch/wire_types.go]: i ended up dropping beacon and kept weight for logging the amount of ignored weight from the missing atxs

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
@dshulyak can you take another look? thank you

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -263,7 +263,7 @@ jobs:
     needs:
       - filter-changes
       - dockerpush
-    timeout-minutes: 60
+    timeout-minutes: 90
```
Comment (dshulyak) [inline .github/workflows/ci.yml:266]: do we still need to increase it, if there is another group?

File: fetch/wire_types.go (hunk 4 above)
Comment (dshulyak) [reply fetch/wire_types.go]: but why to send it? it is not like this number can be trusted

**Review by dshulyak — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 8 above)
Comment (countvonzero) [reply .github/workflows/ci.yml:266]: my understanding is that `systest I` (timeout 50 min) and `systest II` (timeout 30 min) will run sequentially,
hence increasing the overall timeout from 60 -> 90. 
do i understand this incorrectly?

Comment (countvonzero) [reply .github/workflows/ci.yml:266]: actually based on data in https://github.com/spacemeshos/go-spacemesh/actions/runs/3401176463/jobs/5656190035, the two tests are running sequentially.

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**